lib/pbench/server/hierarchy.py

```diff
@@ -0,0 +1,423 @@
+""" Collection and Structure Module for pbench-audit-server
+"""
+
+import os
+import time
+
+from collections import OrderedDict
+
+
+permdict = {
+    "0": "---",
+    "1": "--x",
+    "2": "-w-",
+    "3": "-wx",
+    "4": "r--",
+    "5": "r-x",
+    "6": "rw-",
+    "7": "rwx",
+}
+
+
+def filepermissions(mode):
+    """Convert File permissions from numeric to symbolic.
+    from '755' --> 'rwxr-xr-x' """
+    fperm = ""
+    modebits = oct(mode)[-3:]
+    for bits in modebits:
+        if bits in permdict:
+            fperm += permdict[bits]
+        else:
+            raise ValueError(
+                f"Could not find key '{bits}' in file permissions dictionary"
+            )
+
+    return fperm
+
+
+class Hierarchy(object):
+    """Parent class of Hierarchies"""
+
+    def __init__(self, name, path, config):
+        self.name = name
+        self.path = path
+        self.config = config
+
+    def check_header(self, stat, fp, status):
+        """Check if we need to add header for Results or Incoming Hierarchy"""
+        if stat:
```

Comment:
What does the parameter `stat` mean?

Reply:
it means status. When we calls controller hierarchy then usually the workflow is like we add headers and then print the list of controllers or tarballs and then controller is the one which calls incoming or results hierarchy. This happens when there is list of controllers in controller and list of tarballs in incoming and results. 
Sometimes verifylist(one of controller hierrachy list) consists of controllers which leads to printing of headers but when we move to respective hierarchy(Incoming or results ), tarballs in that controllers are either not valid or fail to pass, which results in output of headers while there is no list to print inside them to prevent this condition stat variable is used and is used in check_header() method.

Reply:
So you have `stat` and `status`, but `stat` means "status"?

Seems like we need to resolve this naming.